README.md
# rClr

## R package for accessing .NET

Accessing the Common Language Runtime (.NET or Mono) from the R statistical software, in-process.

## Prerequisites

As of May 2019, on Windows Clr requires the .NET Framework (4.6.1+, 4.7.2+ recommended). Groundwork towards running on .NET Core started but it is unclear how to embed the .NET Core runtime into R.

## Installing

Releases can be found via the [release tab of the rClr GitHub repository](https://github.com/Open-Systems-Pharmacology/rClr/releases).

### Pre-compiled binaries

You can install pre-compiled rClr for Windows 10 x64, Ubuntu 18, and CentOS 7 via the [release tab of the rClr GitHub repository](https://github.com/Open-Systems-Pharmacology/rClr/releases). Binary packages are available for R 3.6 and 4.x.

### From source

Compiling of the **master** branch from source has been tested under Windows 10 and Linux Ubuntu 18.

#### Windows

On Windows you will need a C# and C and/or Visual C++ compiler. The current releases have been made using the Visual Studio 2019 toolchain. Check the [visual studio download page](https://visualstudio.microsoft.com/downloads/) for options.

It is recommended to build the packages in RStudio.

Following path variables must be present in your environment (see [How to set an environment variable in Windows 10](https://www.onmsft.com/how-to/how-to-set-an-environment-variable-in-windows-10))

- `MSBUILD_EXE_PATH` pointing to MSBuild.exe, e.g. `C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin\MSBuild.exe`
- `VSCOMMON` pointing to the folder where Visual Studio common folder is located, e.g. `C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7`. Alternatively, you can set the path in the file `configure.win'.

#### Linux

Install on Ubuntu 18.04. Mono is supported up to versiono 5.18, so make sure that mono 6 is not installed.

##### GIT
```
sudo apt install git
```

##### NUGET
```
sudo apt install nuget
sudo nuget update -self
```

##### LIBS
```
sudo apt update
sudo apt-get install libcurl4-openssl-dev
sudo apt-get install libssl-dev
sudo apt install libxml2-dev
```

##### MONO

```
sudo apt install gnupg ca-certificates
sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 3FA7E0328081BFF6A14DA29AA6A19B38D3D831EF
echo "deb https://download.mono-project.com/repo/ubuntu stable-bionic/snapshots/5.18 main" | sudo tee /etc/apt/sources.list.d/mono-official-stable.list
sudo apt update
sudo apt install mono-complete
```

Check version with `mono -V`. It should read `5.18.xxx`


##### .NET CORE SDK (For ubuntu 18.04..). 

Install vary based on system. see https://aka.ms/dotnet-download
(e.g. https://docs.microsoft.com/en-us/dotnet/core/install/linux-package-manager-ubuntu-1910)


```
wget -q https://packages.microsoft.com/config/ubuntu/18.04/packages-microsoft-prod.deb -O packages-microsoft-prod.deb
sudo dpkg -i packages-microsoft-prod.deb

sudo add-apt-repository universe
sudo apt-get update
sudo apt-get install apt-transport-https
sudo apt-get update
sudo apt-get install dotnet-sdk-3.1
```

Check version with `dotnet --version`. It should read `3.1.xxx`


##### Install R Packages

Within a R sesssion in a terminal

```
install.packages('devtools')
```

## Getting started

The package contains documentation, code sample and a vignette to get started.

```S
library(rClr)
?rClr
## There is an HTML vignette:
browseVignettes('rClr')
```

OBSOLETE: You will otherwise find some documentation at [https://r2clr.codeplex.com/documentation](https://r2clr.codeplex.com/documentation)

## Feedback and contributions

While this package is sometimes used for the author's paid day job, this is largely a personal endeavour. Support is appreciated in many forms.

* Citations: As of December 2014, [A presentation given at the R user conference 2013](https://publications.csiro.au/rpr/pub?list=ASE&pid=csiro:EP132284&expert=false&sb=RECENT&n=6&rpp=50&page=17&tr=3274&dr=all&csiro.affiliation=B3800). A journal paper will, hmm, "soon" follow.
* Documentation: reporting issues, feature requests or discussion threads as such can be very valuable material if done well.  
* Consulting or contract work is an option that may be arranged.

## Related work

A few packages using rClr are publicly accessible, and may be of interest if you want to build your own package with dependencies on rClr.

* If you are interested in environmental modelling: [RtoTIME](https://github.com/jmp75/RtoTIME) is a package that depends on rClr
* [rsqlserver](https://github.com/agstudy/rsqlserver) is an Sql Server driver database interface (DBI) driver for R
* [A package to access optimization tools on .NET](https://github.com/jmp75/metaheuristics/tree/master/R/pkgs/mh)


NEWS
NEWS/ChangeLog for rClr
--------------------------
0.9.2 2022-28-04
- Compiled with R4.2

0.8-01	2019-04-08	beta
    o	pre-release, test on R 3.5.x which has significant underlying changes. Most of the changes required appeared to be in R.NET though rather than rClr.
	o	Move csproj to netstandard2 format and target.

0.8-0	2015-11-01	beta
    o	Tidy up build process and upgrade to latest recommended practices, notably with testthat

0.7-9	2015-08-30	beta
    o	Update to use R.NET 1.6.5. Following up on a few requests,
	allow for backward compatibility down to
	R 2.15.3 - While all relevant tests work, note that the level
	of testing is not on par with that on R 3.2.x.

0.7-8	2015-08-04	beta
    o	Update to use R.NET 1.6.5. Following up on a few requests,
	allow for backward compatibility down to
	R 2.15.3 - While all relevant tests work, note that the level
	of testing is not on par with that on R 3.2.x.

0.7-4	2015-03-26  beta
    o	Quite a few months and internal versions later, more than
	enough has been done to have a public release. The main change
	under the hood is the use by default of R.NET in preference to C,
	which has opened the conversion of more complicated data structures.

0.7-3	2015-01-15  beta
    o	Make data conversion using R.NET the default.
    o	Feature parity on Linux and Windows

0.7-1	2014-12-11  beta
    o	Quite a few months and internal versions later, more than
	enough has been done to have a public release. The main change
	under the hood is the use by default of R.NET in preference to C,
	which has opened the conversion of more complicated data structures.
	o	One particularly noteworthy feature is the possibility to convert
	.NET objects into R data frames in pure C# (or any other .NET language).
	This is not directly apparent as a feature in rClr itself, but it supports the
	use case. More on these capabilities will come in the form of
	additional documentation in subsequent releases.
	o	Not quite ready functional on Mono, but expected to be at
	parity in upcoming releases.

0.6-1	2014-04-29  beta
    o	Update to use R.NET 1.5.11. This brings lots of performance
	improvements as well as fixes done over the past months to R.NET.
	While not all benefits will be immediately apparent, this is a
	more solid foundation for future rClr versions.

0.5-2	2013-09-21  beta 5
    o	Fix memory leaks for vectors created in R converted to .NET
	arrays. Proper memory release is confirmed for at least numeric,
	integer and character vector types. Part of this exercise,
	rClr has been accepted by Coverity Scan (https://scan.coverity.com).
    I acknowledge Coverity for some helpful insights tracking defects in rClr.
    o	Useful exception handling on MS.NET CLR: print out the details
	of the innermost exception. This is a big improvement on the
	unhelpful HRESULT codes previously reported.
    o	The hosting for the Mono runtime on Windows is included
	in this release. Do note that date-time handling is not
	passing many tests with the Mono CLR.

0.5-1	2013-06-16  beta 4
    o	Adding package documentation using knitr and R-markdown
    o	clrNew now with arguments
    o	Some support for 'enum' types
    o	Date-time handling is now using POSIXct classes instead of Date.
	The temporal resolution is thus much better, however beware that
	POSIXct is effectively a UTC time with a timezone conversion.
	A lot of unit tests are in place around it, but date and time
	zone is a complicated topic.
    o	This release does not include Mono support; unfortunately
	this is not yet usable as it is too unstable. There has
	actually been major progress, but some new issues still
	emerged.


R/rClr-package.r

## Note to self: see the roxygen vignette for examples

#' Accessing the Common Language Runtime (.NET/Mono) from R
#' 
#' \tabular{ll}{
#' Package: \tab rClr \cr
#' Type: \tab Package \cr
#' Version: \tab 0.9.0 \cr
#' Notes: \tab Works with R 4.0.x series. Credits to Pete Thompson https://github.com/pete-thompson for some of the changes. Uses R.NET 1.9.0. \cr
#' Date: \tab 2020-10-05\cr
#' License: \tab LGPL 3\cr
#' }
#'
#' rClr is a package for low-level access from R to a Common Language Runtime (CLR). The supported CLR implementations are Microsoft '.NET'
#' framework on Windows and Mono on several platforms, currently Windows and at least Debian Linux. rClr has been used in scientific modelling work since
#' 2013-06. Running it on Mono has been relatively experimental, but as of November 2014 several factors have enabled very near functional parity with the Windows/.NET.
#'
#' \tabular{lll}{
#' Version \tab Date \tab Notes \cr
#' 0.8.3: \tab 2019-05-09 \tab Works with R 3.5.x series. .NET code targets netstandard2.0. \cr
#' 0.8.2: \tab 2019-04-09 \tab preview release - work on R 3.5.x, which has significant changes under the hood. \cr
#' 0.8.1: \tab 2019-04-08 \tab preview release - work on R 3.5.x, which has significant changes under the hood. \cr
#' 0.8.0: \tab 2015-11-01 \tab Tidy up build process and upgrade to latest recommended practices, notably with testthat. \cr
#' 0.7-9: \tab 2015-08-30 \tab allow compilation against boehm and SGen versions of mono lib. minor build improvements. \cr
#' 0.7-8: \tab 2015-08-04 \tab Update to use R.NET 1.6.5. Allow for backward compatibility down to R 2.15.3 - While all relevant tests work, note that the level of testing is not on par with tests on R 3.2.x. \cr
#' 0.7-7: \tab 2015-06-18 \tab Update to use R.NET 1.6.4. Functionally unchanged otherwise. Not a public release \cr
#' 0.7-6: \tab 2015-05-07 \tab update the R Markdown vignettes to work with the latest version of knitr. Not a public release \cr
#' 0.7-5: \tab 2015-04-06 \tab Internal release for work project. Compile against mono sgen rather than boehm. \cr
#' }
#'
#' I gratefully acknowledge Kosei Abe and others for their work on the dependency R.NET (see https://github.com/rdotnet/rdotnet/graphs/contributors)
#'
#' @name rClr-package
#' @aliases rClr
#' @docType package
#' @title R accessing .NET/Mono 
#' @author Jean-Michel Perraud \email{jean-michel.perraud_at_csiro.au}
#' @keywords package CLR Mono .NET
NULL



